Clean up miscellaneous code (mostly cache-related)

showell opened the pull request.

timabbott commented on zerver/actions/realm_emoji.py:
```diff
@@ -49,7 +57,7 @@ def check_add_realm_emoji(
     realm_emoji.save(update_fields=["file_name", "is_animated"])
 
-    realm_emoji_dict = realm.get_emoji()
+    realm_emoji_dict = get_realm_emoji_by_id(realm.id)
```
Can we do `get_all_realm_emoji_by_id(realm_id)` or `get_custom_emoji_for_realm` for this name? Read this, I think we're looking up a single emoji by `realm_emoji.id`, sigh.

showell replied:
I'll try out some better names.  I'm probably gonna squash the prep commit here into the commit that unifies the cache, since you've already taken a review pass, and since there's no really good intermediate state of the refactoring.

timabbott commented on zerver/models.py:
```diff
@@ -1196,21 +1186,15 @@ def get_realm_emoji_dicts(realm_id: int, only_active_emojis: bool = False) -> Di
+def get_realm_active_emoji(realm_id: int) -> Dict[str, EmojiInfo]:
+    # It's important to use the cached version here.
+    realm_emojis = get_realm_emoji_by_id(realm_id)
+    return {row["name"]: row for row in realm_emojis.values() if not row["deactivated"]}
```
I think in my ideal world, we'd have a single function called by other modules, `get_realm_emoji(realm_id: int, *, include_deactivated: bool)`, so that one can't call `get_realm_emoji_by_id` and accidentally not realize you're incorrectly including deactivated custom emoji that should be ignored ... and that function would, use a single cache that just gets all the custom emoji by the realm and just filters internally the deactivated custom emoji to pass back to its caller.

Thoughts?

showell replied:
The hindrance to this suggestion is that when you're dealing with active emojis you want "name" as the key, whereas the other function uses "id" as the key.  I kinda hate boolean args (even when they are explicit, as in your suggestion), and I hate even more to return some kind of Union type.

I'll address this better names.

timabbott commented on zerver/lib/emoji.py:
```diff
@@ -62,17 +62,29 @@ def translate_emoticons(text: str) -> str:
+    reaction_type: str
+
+
+def get_emoji_data(realm_id: int, emoji_name: str) -> EmojiData:
```
How about `get_emoji_data_for_name`?

showell replied:
I pretty strongly prefer brevity here.  We don't have any similar functions where the distinction would be confusing.  The original name of this function was overly precise, which led to it eventually being wrong.  I'd rather be brief and vague than overly precise and wrong.

The whole emoji system is kind of confusing with the names and codes and ids, but I would rather address this with big scary comments (like we already have in a couple of the views) than with naming.  And, of course, whoever takes the next pass at this can always improve the name based on newer knowledge or requirements.

showell commented:
This is ready for review again.  I addressed your comments above.

New stuff this round:
* squashed the two commits that consolidate the cache
* improved some naming decisions
* cleaned up some repetitive test code

Feel free to tweak the names of the three new function names without a review round trip.  I don't think any of my naming decisions are clearly perfect, but I also think there are no perfect solutions here.  I mostly want to fix the structural issues here, and we can always re-visit some of the names later.

timabbott commented:
Looks good, merged, thanks @showell!
